File: scratch/weighted_2026-09-15/classes.py
"""E_P 2^{-r(P)} = (#swap classes)/|P| for c-confined prefixes of length j0 at the top shell (pairs (2k,2k+1)).
This is the full-group value of the class-averaged swap-cube L^2 bound; its rate -log2(.)/(2 j0) is the
amplitude rate the swap-cube route delivers if interval averages match full-group averages."""
import math
al = math.log2(3)
def run(c, J, gap=0):
    b = [math.floor(c + j * al) for j in range(J + 3)]; sig = b[J] - gap
    cnt = [dict() for _ in range(J + 1)]; g = [dict() for _ in range(J + 1)]
    cnt[J][sig] = 1.0; g[J][sig] = 1.0
    i = J
    if J % 2 == 1:
        i = J - 1
        for S in range(0, b[i] + 1):
            v = sum(cnt[J].get(S + d, 0) for d in range(1, b[J] - S + 1))
            if v: cnt[i][S] = v; g[i][S] = v
    i -= 2
    while i >= 0:
        for S in range(0, b[i] + 1):
            vc = vg = 0.0
            for d in range(1, b[i + 1] - S + 1):
                for e in range(1, b[i + 2] - S - d + 1):
                    x = cnt[i + 2].get(S + d + e, 0)
                    if not x: continue
                    vc += x
                    sw = d != e and S + e <= b[i + 1]
                    vg += g[i + 2][S + d + e] * (0.5 if sw else 1.0)
            if vc: cnt[i][S] = vc; g[i][S] = vg
        i -= 2
    P = cnt[0][0]; C = g[0][0]
    return P, C
need = 0.0911
for c, J in [(0, 30), (0, 60), (0, 100), (0, 150), (0, 200), (0, 300), (0, 400), (1, 200), (2, 200)]:
    P, C = run(c, J)
    rate = -math.log2(C / P) / (2 * J)
    print(f"c={c} j0={J}: log2|P|={math.log2(P):.1f}  log2 #classes={math.log2(C):.1f}  E 2^-r = 2^{math.log2(C/P):.1f}  -> amplitude rate {rate:.4f} bits/step (need {need})")

File: scratch/weighted_2026-09-15/classes.c
// E_P 2^{-r(P)} = #swap classes / |P| (pairs (2k,2k+1), swappable: d != e and both orders admissible)
// = full-group value of the class-averaged swap-cube L^2 bound -> ideal swap-cube amplitude rate.
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
static int Kb[2048];
int main(void) {
  const long double A = 1.584962500721156181453738943947816508759814407692481060455L;
  int Js[] = {30, 60, 100, 150, 200, 300, 400, 600}; int cs[] = {0, 1, 2};
  for (int ci = 0; ci < 3; ci++) for (int jj = 0; jj < 8; jj++) {
    int c = cs[ci], J = Js[jj]; if (c > 0 && J != 200) continue;
    for (int j = 0; j < 2048; j++) Kb[j] = (int)floorl(c + j * A);
    int sg = Kb[J], D = sg + 2;
    double *cn = calloc(D, 8), *gn = calloc(D, 8), *cn2 = calloc(D, 8), *gn2 = calloc(D, 8);
    cn[sg] = 1; gn[sg] = 1; int i = J; double logscale = 0;
    if (J % 2) { i = J - 1; for (int S = 0; S <= Kb[i]; S++) { double v = 0; for (int d = 1; S + d <= Kb[J]; d++) v += cn[S + d]; cn2[S] = v; gn2[S] = v; }
      memcpy(cn, cn2, 8 * D); memcpy(gn, gn2, 8 * D); }
    for (i = i - 2; i >= 0; i -= 2) {
      memset(cn2, 0, 8 * D); memset(gn2, 0, 8 * D);
      for (int S = 0; S <= Kb[i]; S++) { double vc = 0, vg = 0;
        for (int d = 1; S + d <= Kb[i + 1]; d++) for (int e = 1; S + d + e <= Kb[i + 2]; e++) {
          double x = cn[S + d + e]; if (x == 0) continue; vc += x;
          int sw = (d != e) && (S + e <= Kb[i + 1]); vg += gn[S + d + e] * (sw ? 0.5 : 1.0); }
        cn2[S] = vc; gn2[S] = vg; }
      double mx = 0; for (int S = 0; S < D; S++) if (cn2[S] > mx) mx = cn2[S];
      for (int S = 0; S < D; S++) { cn[S] = cn2[S] / mx; gn[S] = gn2[S] / mx; } logscale += log2(mx);
    }
    double ratio = gn[0] / cn[0];
    printf("c=%d j0=%d: log2|P|=%.1f  E_P 2^-r = 2^%.2f  -> ideal swap-cube amplitude rate %.4f bits/step\n",
           c, J, logscale + log2(cn[0]), log2(ratio), -log2(ratio) / (2.0 * J));
    free(cn); free(gn); free(cn2); free(gn2);
  }
  return 0;
}
